File: packages/soulseek_protocol/lib/src/connection/reconnection_manager.dart
import 'dart:async';
import 'dart:math';

import 'socket_manager.dart';

class ReconnectionManager {
  final SocketTransport _socketManager;
  final ReconnectionConfig _config;

  bool _isRunning = false;
  int _attempt = 0;
  Timer? _timer;
  String? _host;
  int? _port;

  final _stateController = StreamController<ReconnectionState>.broadcast();

  ReconnectionManager(this._socketManager, {ReconnectionConfig? config})
      : _config = config ?? ReconnectionConfig();

  Stream<ReconnectionState> get stateChanges => _stateController.stream;
  ReconnectionState get state => ReconnectionState(
    isRunning: _isRunning,
    attempt: _attempt,
    nextDelay: _calculateDelay(),
    lastError: null,
  );

  void start(String host, int port) {
    _host = host;
    _port = port;
    _isRunning = true;
    _attempt = 0;
    _socketManager.stateChanges.listen(_onSocketStateChange);
    _tryReconnect();
  }

  void stop() {
    _isRunning = false;
    _timer?.cancel();
    _timer = null;
    _attempt = 0;
    _emitState();
  }

  void reset() {
    _attempt = 0;
    _timer?.cancel();
    _emitState();
  }

  void _onSocketStateChange(SocketStateChanged event) {
    if (event.state == SocketState.connected) {
      _isRunning = false;
      _attempt = 0;
      _timer?.cancel();
      _emitState();
    } else if (event.state == SocketState.disconnected && _isRunning) {
      _scheduleReconnect(event.errorMessage);
    }
  }

  void _scheduleReconnect([String? errorMessage]) {
    _timer?.cancel();
    final delay = _calculateDelay();
    _attempt++;
    _emitState(errorMessage: errorMessage);

    _timer = Timer(delay, _tryReconnect);
  }

  void _tryReconnect() {
    if (!_isRunning || _host == null || _port == null) return;

    if (_config.maxAttempts >= 0 && _attempt >= _config.maxAttempts) {
      _isRunning = false;
      _emitState(isFinal: true);
      return;
    }

    _socketManager.connect(_host!, _port!).catchError((_) {});
  }

  Duration _calculateDelay() {
    if (_attempt == 0) return Duration.zero;
    final baseMs = _config.baseDelay.inMilliseconds;
    final maxMs = _config.maxDelay.inMilliseconds;
    final delay = min(baseMs * pow(_config.multiplier, _attempt - 1).toInt(), maxMs);
    final jitter = _config.jitterMs > 0
        ? Random().nextInt(_config.jitterMs) - (_config.jitterMs ~/ 2)
        : 0;
    return Duration(milliseconds: delay + jitter);
  }

  void _emitState({String? errorMessage, bool isFinal = false}) {
    if (_stateController.isClosed) return;
    _stateController.add(ReconnectionState(
      isRunning: _isRunning,
      attempt: _attempt,
      nextDelay: _calculateDelay(),
      lastError: errorMessage,
      isFinal: isFinal,
    ));
  }

  void dispose() {
    stop();
    _stateController.close();
  }
}

class ReconnectionConfig {
  final Duration baseDelay;
  final Duration maxDelay;
  final double multiplier;
  final int maxAttempts;
  final int jitterMs;

  const ReconnectionConfig({
    this.baseDelay = const Duration(seconds: 1),
    this.maxDelay = const Duration(seconds: 60),
    this.multiplier = 2.0,
    this.maxAttempts = -1, // -1 means unlimited
    this.jitterMs = 1000,
  });
}

class ReconnectionState {
  final bool isRunning;
  final int attempt;
  final Duration nextDelay;
  final String? lastError;
  final bool isFinal;

  const ReconnectionState({
    required this.isRunning,
    required this.attempt,
    required this.nextDelay,
    this.lastError,
    this.isFinal = false,
  });
}

File: Soulseek.Protocol/Connection/ReconnectionManager.cs
namespace Soulseek.Protocol.Connection;

public class ReconnectionConfig
{
    public TimeSpan BaseDelay { get; init; } = TimeSpan.FromSeconds(1);
    public TimeSpan MaxDelay { get; init; } = TimeSpan.FromSeconds(60);
    public double Multiplier { get; init; } = 2.0;
    public int MaxAttempts { get; init; } = -1;
    public int JitterMs { get; init; } = 1000;
}

public record ReconnectionState(
    bool IsRunning,
    int Attempt,
    TimeSpan NextDelay,
    string? LastError = null,
    bool IsFinal = false
);

public class ReconnectionManager
{
    private readonly ISocketTransport _socketTransport;
    private readonly ReconnectionConfig _config;

    private bool _isRunning;
    private int _attempt;
    private Timer? _timer;
    private string? _host;
    private int? _port;
    private IDisposable? _stateSub;

    private readonly Subject<ReconnectionState> _stateSubject = new();

    public ReconnectionManager(ISocketTransport socketTransport, ReconnectionConfig? config = null)
    {
        _socketTransport = socketTransport;
        _config = config ?? new ReconnectionConfig();
    }

    public IObservable<ReconnectionState> StateChanges => _stateSubject;
    public ReconnectionState State => new(_isRunning, _attempt, CalculateDelay());

    public void Start(string host, int port)
    {
        _host = host;
        _port = port;
        _isRunning = true;
        _attempt = 0;
        _stateSub = _socketTransport.StateChanges.Subscribe(OnSocketStateChange);
        TryReconnect();
    }

    public void Stop()
    {
        _isRunning = false;
        _timer?.Dispose();
        _timer = null;
        _attempt = 0;
        _stateSub?.Dispose();
        _stateSub = null;
        EmitState();
    }

    public void Reset()
    {
        _attempt = 0;
        _timer?.Dispose();
        EmitState();
    }

    private void OnSocketStateChange(SocketStateChanged eventArgs)
    {
        if (eventArgs.State == SocketState.Connected)
        {
            _isRunning = false;
            _attempt = 0;
            _timer?.Dispose();
            EmitState();
        }
        else if (eventArgs.State == SocketState.Disconnected && _isRunning)
        {
            ScheduleReconnect(eventArgs.ErrorMessage);
        }
    }

    private void ScheduleReconnect(string? errorMessage = null)
    {
        _timer?.Dispose();
        var delay = CalculateDelay();
        _attempt++;
        EmitState(errorMessage: errorMessage);

        _timer = new Timer(_ => TryReconnect(), null, delay, Timeout.InfiniteTimeSpan);
    }

    private void TryReconnect()
    {
        if (!_isRunning || _host == null || _port == null) return;

        if (_config.MaxAttempts >= 0 && _attempt >= _config.MaxAttempts)
        {
            _isRunning = false;
            EmitState(isFinal: true);
            return;
        }

        _ = _socketTransport.Connect(_host!, _port!.Value);
    }

    private TimeSpan CalculateDelay()
    {
        if (_attempt == 0) return TimeSpan.Zero;
        var baseMs = _config.BaseDelay.TotalMilliseconds;
        var maxMs = _config.MaxDelay.TotalMilliseconds;
        var delay = Math.Min(baseMs * Math.Pow(_config.Multiplier, _attempt - 1), maxMs);
        var jitter = _config.JitterMs > 0
            ? Random.Shared.Next(_config.JitterMs) - (_config.JitterMs / 2)
            : 0;
        return TimeSpan.FromMilliseconds(delay + jitter);
    }

    private void EmitState(string? errorMessage = null, bool isFinal = false)
    {
        _stateSubject.OnNext(new ReconnectionState(
            _isRunning, _attempt, CalculateDelay(), errorMessage, isFinal));
    }

    public void Dispose()
    {
        Stop();
        _stateSubject.Dispose();
    }
}